File: holoken-core/src/main/java/com/srlee/dlx/MathDokuDLX.java
package com.srlee.dlx;

import com.holokenmod.Grid;
import com.holokenmod.GridCage;
import com.holokenmod.creation.GridSingleCageCreator;
import com.holokenmod.options.DigitSetting;
import com.holokenmod.options.GameVariant;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

public class MathDokuDLX extends DLX {
	private final static Logger LOGGER = LoggerFactory.getLogger(MathDokuDLX.class);
	
	public MathDokuDLX(final Grid grid) {
		
		// Number of columns = number of constraints =
		//		BOARD * BOARD (for columns) +
		//		BOARD * BOARD (for rows)	+
		//		Num cages (each cage has to be filled once and only once)
		// Number of rows = number of "moves" =
		//		Sum of all the possible cage combinations
		// Number of nodes = sum of each move:
		//      num_cells column constraints +
		//      num_cells row constraints +
		//      1 (cage constraint)
		int total_nodes = 0;
		
		final Collection<GridSingleCageCreator> creators = new ArrayList<>();
		
		for (final GridCage cage : grid.getCages()) {
			creators.add(new GridSingleCageCreator(grid, cage));
		}
		
		for (final GridSingleCageCreator creator : creators) {
			total_nodes += creator.getPossibleNums().size() * (2 * creator.getNumberOfCells() + 1);
		}
		init(2 * grid.getGridSize().getSurfaceArea() + creators.size(), total_nodes);
		
		int constraint_num;
		int move_idx = 0;
		
		DigitSetting digitSetting = GameVariant.getInstance().getDigitSetting();
		
		for (final GridSingleCageCreator creator : creators) {
			for (final int[] onemove : creator.getPossibleNums()) {
				LOGGER.info("cage " + creator.getCage() + " - " + Arrays.toString(onemove));
				
				for (int i = 0; i < onemove.length; i++) {
					int numberToTestIndex = digitSetting.indexOf(onemove[i]);
					
					constraint_num = grid.getGridSize().getWidth() * numberToTestIndex + creator.getCell(i)
							.getColumn() + 1;
					addNode(constraint_num, move_idx);    // Column constraint
					constraint_num = grid.getGridSize().getSurfaceArea() + grid.getGridSize().getWidth() * numberToTestIndex + creator.getCell(i)
							.getRow() + 1;
					addNode(constraint_num, move_idx);    // Row constraint
				}
				constraint_num = 2 * grid.getGridSize().getSurfaceArea() + creator.getId() + 1;
				addNode(constraint_num, move_idx);    // Cage constraint
				move_idx++;
			}
		}
	}
}

File: holoken-core/src/main/kotlin/com/srlee/dlx/MathDokuDLX.kt
package com.srlee.dlx

import com.holokenmod.creation.cage.GridSingleCageCreator
import com.holokenmod.grid.Grid

class MathDokuDLX(grid: Grid) {
    private var dlx: DLX

    init {

        // Number of columns = number of constraints =
        // 		BOARD * BOARD (for columns) +
        // 		BOARD * BOARD (for rows)	+
        // 		Num cages (each cage has to be filled once and only once)
        // Number of rows = number of "moves" =
        // 		Sum of all the possible cage combinations
        // Number of nodes = sum of each move:
        //      num_cells column constraints +
        //      num_cells row constraints +
        //      1 (cage constraint)
        var numberOfNodes = 0
        val creators: MutableCollection<GridSingleCageCreator> = ArrayList()
        for (cage in grid.cages) {
            creators.add(GridSingleCageCreator(grid.variant, cage))
        }
        for (creator in creators) {
            numberOfNodes += creator.possibleNums.size * (2 * creator.numberOfCells + 1)
        }

        dlx = DLX(2 * grid.gridSize.surfaceArea + creators.size, numberOfNodes)

        var currentCombination = 0
        val digitSetting = grid.options.digitSetting
        for (creator in creators) {
            for (possibleCageCombination in creator.possibleNums) {
                for (i in possibleCageCombination.indices) {
                    val indexOfDigit = digitSetting.indexOf(possibleCageCombination[i])

                    val columnConstraint =
                        grid.gridSize.width * indexOfDigit + creator.getCell(i).column
                    val rowConstraint =
                        grid.gridSize.surfaceArea + grid.gridSize.width * indexOfDigit + creator.getCell(
                            i
                        ).row

                    dlx.addNode(columnConstraint, currentCombination)
                    dlx.addNode(rowConstraint, currentCombination)
                }

                val cageConstraint = 2 * grid.gridSize.surfaceArea + creator.id

                dlx.addNode(cageConstraint, currentCombination)

                currentCombination++
            }
        }
    }

    fun solve(type: DLX.SolveType): Int {
        return dlx.solve(type)
    }
}
